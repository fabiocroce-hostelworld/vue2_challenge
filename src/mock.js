import { clavaAPI } from "@/api/clavaAPI"

clavaAPI.api.get = jest.fn()

export const mockAxiosGet = clavaAPI.api.get
