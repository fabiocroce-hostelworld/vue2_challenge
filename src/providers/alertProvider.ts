import { provide, inject, reactive, computed, ComputedRef } from "@vue/composition-api"

export enum alertTypes {
  SUCCESS = "success",
  ERROR = "error",
  INFO = "info",
  WARNING = "warning"
}

interface Alert {
  type: alertTypes
  message: string
  timerId?: number | undefined
}

export interface AlertProvider {
  addErrorAlert: (message: string) => void
  addSuccessAlert: (message: string) => void
  clearAlerts: () => void
  clearAlert: (idx: number) => void
  alerts: ComputedRef<Alert[]>
}

type alertConfig = { timeout?: number }

// Store for alert messages
export const setupAlertProvider = (config: alertConfig = {}): NonNullable<AlertProvider> => {
  const alertsArray = reactive<{ alerts: Array<Alert> }>({ alerts: [] })

  const clearByTimerId = (removeTimerId: number) => {
    alertsArray.alerts = alertsArray.alerts.filter(({ timerId }) => timerId !== removeTimerId)
  }

  const addAlert = ({ message, type }: Alert) => {
    if (config.timeout) {
      const timerId = setTimeout(() => {
        clearByTimerId(timerId)
      }, config.timeout)

      alertsArray.alerts.push({ message, type, timerId })
      return
    }
    alertsArray.alerts.push({ message, type })
  }

  const addErrorAlert = (message: string) => addAlert({ message, type: alertTypes.ERROR })
  const addSuccessAlert = (message: string) => addAlert({ message, type: alertTypes.SUCCESS })

  const clearAlerts = () => void (alertsArray.alerts.length = 0)
  const clearAlert = (removeIdx: number) => {
    alertsArray.alerts = alertsArray.alerts.filter((_, idx) => removeIdx !== idx)
  }
  const alerts = computed(() => alertsArray.alerts)

  return {
    addErrorAlert,
    addSuccessAlert,
    clearAlerts,
    clearAlert,
    alerts
  }
}

export const alertRef = Symbol("alert")

export const provideAlert = /* istanbul ignore next */ (config: alertConfig = {}) =>
  provide(alertRef, setupAlertProvider(config))
export const useAlert = () => inject<AlertProvider>(alertRef)
