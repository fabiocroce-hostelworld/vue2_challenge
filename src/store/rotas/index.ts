import { getRotaServices } from "@/api/clavaAPI"
import { IRotas } from "@/api/clavaAPI/rota/RotasResponseDTO"
import { IRotasList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_ROTAS: "GET_ROTAS"
  },
  mutations: {
    STORE_ROTAS: "STORE_ROTAS",
    STORE_ROTA: "STORE_ROTA"
  },
  actions: {
    FETCH_ROTA: "fetchRota",
    GENERATE_ROTA: "generateRota"
  }
}

export const rotasTypes = addNamespacePrefix(types, "rotas")

type RotasState = {
  rotas: IRotasList
}

export default {
  namespaced: true,
  state: (): RotasState => ({
    rotas: [] as IRotasList
  }),
  getters: {
    [types.getters.GET_ROTAS](state: RotasState) {
      return state.rotas
    }
  },
  actions: {
    [types.actions.FETCH_ROTA]: async ({ commit }: { commit: Function }, rotaId: number) => {
      const rota = await getRotaServices().fetchRota(rotaId)

      commit(types.mutations.STORE_ROTA, rota)
    },
    [types.actions.GENERATE_ROTA]: async ({ dispatch }: { dispatch: Function }, date: string) => {
      const rotaId = await getRotaServices().generateRota(date)

      await dispatch(types.actions.FETCH_ROTA, rotaId)
    }
  },
  mutations: {
    [types.mutations.STORE_ROTAS]: (state: RotasState, rotas: IRotasList) => {
      state.rotas = rotas
    },
    [types.mutations.STORE_ROTA]: (state: RotasState, rota: IRotas) => {
      state.rotas.push(rota)
    }
  }
}
