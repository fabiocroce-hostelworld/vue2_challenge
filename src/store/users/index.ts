import { IUsersList } from "@/api/clavaAPI/rotasUsers/RotasResponseDTO"
import { addNamespacePrefix } from "@/utils/vuexHelpers"

const types = {
  getters: {
    GET_USERS: "getUsers",
    GET_MAPPED_USERS: "getMappedUsers",
    GET_USER_NAME: "getUserName"
  },
  mutations: {
    STORE_USERS: "STORE_USERS"
  },
  actions: {
    FETCH_USERS: "fetchUsers"
  }
}

export const usersTypes = addNamespacePrefix(types, "users")

type UsersState = {
  users: IUsersList
}

export default {
  namespaced: true,
  state: (): UsersState => ({
    users: [] as IUsersList
  }),
  getters: {
    [types.getters.GET_USERS](state: UsersState) {
      return state.users
    },
    [types.getters.GET_MAPPED_USERS](state: UsersState) {
      return Object.assign({}, ...state.users.map(user => ({ [user.userId]: user.user })))
    },
    [types.getters.GET_USER_NAME](_: UsersState, getters: { [x: string]: IUsersList }) {
      return (userId: number) => getters[types.getters.GET_MAPPED_USERS][userId]
    }
  },
  mutations: {
    [types.mutations.STORE_USERS]: (state: UsersState, users: IUsersList) => {
      state.users = users
    }
  }
}
