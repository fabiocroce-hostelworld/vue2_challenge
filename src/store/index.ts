import { getRotasUsersServices } from "@/api/clavaAPI"
import Vue from "vue"
import Vuex, { Store } from "vuex"
import rotas, { rotasTypes } from "@/store/rotas"
import users, { usersTypes } from "@/store/users"

Vue.use(Vuex)

export const rootTypes = {
  actions: {
    FETCH_ROTAS_AND_USERS: "fetchRotasAndUsers"
  }
}

export { rotasTypes, usersTypes }

export const storeConfig = {
  modules: {
    rotas,
    users
  },
  actions: {
    [rootTypes.actions.FETCH_ROTAS_AND_USERS]: async ({ commit }: { commit: Function }) => {
      const { users, rotas } = await getRotasUsersServices().fetchUsersAndRotas()

      commit(rotasTypes.mutations.STORE_ROTAS, rotas)
      commit(usersTypes.mutations.STORE_USERS, users)
    }
  }
}

export default new Store(storeConfig)
