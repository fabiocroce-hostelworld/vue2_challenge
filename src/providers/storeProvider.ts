import { provide, inject, computed, ComputedRef } from "@vue/composition-api"
import store from "@/store"

export interface StoreProvider {
  store: typeof store
  reactiveGetter: <T>(name: string) => ComputedRef<T>
}

// Vuex adapter for composition API
export const setupStoreProvider = (localStore: typeof store): StoreProvider => ({
  store: localStore,
  reactiveGetter: <T>(getterName: string): ComputedRef<T> => {
    return computed(() => localStore.getters[getterName])
  }
})

export const storeRef = Symbol("vuex-store")
export const provideStore = /* istanbul ignore next */ () => provide(storeRef, setupStoreProvider(store))
export const useStore = () => inject<StoreProvider>(storeRef)
