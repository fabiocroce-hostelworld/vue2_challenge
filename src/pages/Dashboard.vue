<template>
  <v-app>
    <h1 class="primary--text">DASHBOARD</h1>
    <alert v-if="alerts.length" class="alerts" />
    <rotas-container v-if="isSuccess" class="rotas-container" />
    <v-progress-circular v-else :size="70" color="primary" indeterminate />
  </v-app>
</template>

<script lang="ts">
import RotasContainer from "@/components/dashboard/RotasContainer.vue"
import Alert from "@/components/helpers/Alert.vue"
import { rootTypes } from "@/store"
import { useFetch } from "@/hooks/complex/useFetch"
import { useStore } from "@/providers/storeProvider"
import { useAlert } from "@/providers/alertProvider"
import { defineComponent, onMounted } from "@vue/composition-api"

export default defineComponent({
  name: "Dashboard",
  components: {
    RotasContainer,
    Alert
  },
  setup() {
    const { store } = useStore()!
    const { addSuccessAlert, addErrorAlert, alerts } = useAlert()!
    const { fetchData, isSuccess } = useFetch()

    onMounted(() => {
      fetchData(() => store.dispatch(rootTypes.actions.FETCH_ROTAS_AND_USERS))
        .then(() => void addSuccessAlert("Rotas have been fetched successfully."))
        .catch(() => void addErrorAlert("Could not fetch rotas."))
    })

    return {
      isSuccess,
      alerts: alerts.value
    }
  }
})
</script>

<style>
#app {
  font-family: Avenir, Helvetica, Arial, sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  margin-top: 60px;
}
</style>
