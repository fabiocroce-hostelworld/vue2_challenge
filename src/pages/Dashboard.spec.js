import Vuex from "vuex"
import { shallowMount, createLocalVue } from "@vue/test-utils"
import { storeConfig } from "@/store"
import Dashboard from "@/pages/Dashboard"
import rotaUsersMock from "@/api/clavaAPI/rotasUsers/rotasResponse.mock"
import { setupStoreProvider, storeRef } from "@/providers/storeProvider"
import { setupAlertProvider, alertRef } from "@/providers/alertProvider"
import { mockAxiosGet } from "@/mock"
import flushPromises from "flush-promises"

const alertProvider = setupAlertProvider({})
alertProvider.addErrorAlert = jest.fn()
alertProvider.addSuccessAlert = jest.fn()

describe("Dashboard", () => {
  let input
  let localVue
  let store

  beforeEach(() => {
    localVue = createLocalVue()
    store = new Vuex.Store(storeConfig)

    input = {
      localVue,
      provide: {
        [storeRef]: setupStoreProvider(store),
        [alertRef]: alertProvider
      }
    }
  })

  it("Snapshot test - loading", () => {
    const wrapper = shallowMount(Dashboard, input)

    expect(wrapper.findAll(".rotas-container").length).toBe(0)
    expect(wrapper.findAll("v-progress-circular-stub").length).toBe(1)
    expect(wrapper.findAll(".alerts").length).toBe(0)

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("Snapshot test - loaded", async () => {
    mockAxiosGet.mockResolvedValueOnce({ data: rotaUsersMock, status: 200 })
    const wrapper = shallowMount(Dashboard, input)

    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("Snapshot test - failed", async () => {
    mockAxiosGet.mockRejectedValue({ message: "error" })
    const wrapper = shallowMount(Dashboard, input)

    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it("test if when data is fetched if the rotas container and alert appears", async () => {
    mockAxiosGet.mockResolvedValueOnce({ data: rotaUsersMock, status: 200 })
    const wrapper = shallowMount(Dashboard, input)

    await flushPromises()

    expect(alertProvider.addSuccessAlert).toHaveBeenCalled()

    expect(wrapper.findAll(".rotas-container").length).toBe(1)
    expect(wrapper.findAll("v-progress-circular-stub").length).toBe(0)
  })

  it("test if API call fails", async () => {
    mockAxiosGet.mockRejectedValue({ message: "error" })
    const wrapper = shallowMount(Dashboard, input)

    await flushPromises()

    expect(wrapper.findAll(".rotas-container").length).toBe(0)
    expect(wrapper.findAll("v-progress-circular-stub").length).toBe(1)
    expect(alertProvider.addErrorAlert).toHaveBeenCalled()
  })
})
