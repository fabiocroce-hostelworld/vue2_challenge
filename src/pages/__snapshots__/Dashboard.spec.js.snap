// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard Snapshot test - failed 1`] = `
<v-app-stub id="app">
  <h1 class="primary--text">DASHBOARD</h1>
  <!---->
  <v-progress-circular-stub color="primary" indeterminate="true" rotate="0" size="70" width="4" value="0"></v-progress-circular-stub>
</v-app-stub>
`;

exports[`Dashboard Snapshot test - loaded 1`] = `
<v-app-stub id="app">
  <h1 class="primary--text">DASHBOARD</h1>
  <!---->
  <rotas-container-stub class="rotas-container"></rotas-container-stub>
</v-app-stub>
`;

exports[`Dashboard Snapshot test - loading 1`] = `
<v-app-stub id="app">
  <h1 class="primary--text">DASHBOARD</h1>
  <!---->
  <v-progress-circular-stub color="primary" indeterminate="true" rotate="0" size="70" width="4" value="0"></v-progress-circular-stub>
</v-app-stub>
`;
